### ISSUE TITLE
where can i get the swe solver

### ISSUE THREAD

**Issue by ZhaoMY93** [obdwinston/Incompressible-Flow#1]
Hello, I remember that you used to have a solver for shallow water in your library, but I can't seem to find it now. Is it still accessible? If so, how can I access it specifically?

**Comment by obdwinston** [comment]
Hi Zhao, please see attached for the shallow flow solver. Let me know if
you need any more details!

Regards
Winston

**Comment by obdwinston** [comment]
[Shallow-Flow.zip](https://github.com/user-attachments/files/18222257/Shallow-Flow.zip)

See attached. Let me know if you need any more details!

**Comment by ZhaoMY93** [comment]
Dear,
Thank you very much for your reply! Thank you so much! May I ask you another question?
I remember that in May I spent half a month primarily studying your shallow water code. At that time, I discovered a small issue, but I couldn't resolve it. I found that it runs normally in your example, but when I replace it with other larger terrains, the model diverges after running for a while (the stable running time remains relatively short). I have carefully debugged it and found that even when I set the time step to be very small, as time accumulates, the model still fails after a few seconds due to too large flux iterations. I meticulously checked each step and found that the accumulation of this error leading to divergence is likely mainly due to the calculation of fluxes. On some grids, the flux becomes increasingly large, reaching a very large magnitude, and then fails.
What confuses me is that in the MONAI example, it can indeed run stably for a long time. I have also read literature on the approach, and I believe that the numerical methods in the code itself are correct. One major difference between the MONAI example and the terrain data I replaced might be that the grid subdivision in my real terrain data is not as dense as in the example. This is because the terrain in the example is scaled down, with actual distances of several kilometers reduced to a few meters. Therefore, each grid represents a relatively small area. However, in the terrain data I replaced, the number of grids used is roughly the same as in MONAI, but due to the larger actual terrain, each grid represents a larger area. Do you think this could be the reason? Or have you run this model code on larger real terrains?
Yours sincerely,


Ray


发件人：Winston Ong ***@***.***>
发送日期：2024-12-22 22:50:01
收件人：obdwinston/Incompressible-Flow ***@***.***>
抄送人：ZhaoMY93 ***@***.***>,Author ***@***.***>
主题：Re: [obdwinston/Incompressible-Flow] where can i get the swe solver (Issue #1)

Hi Zhao, please see attached for the shallow flow solver. Let me know if
 you need any more details!

 Regards
 Winston

**Comment by obdwinston** [comment]
Yes, it is very possible that the scale of your terrain may cause the
solution to diverge. Based on personal experience, that may often be the
case (assuming code has no errors), not just for this shallow flow solver.
Hence, it's important that the scales are appropriate. Usually this
corresponds to working with non-dimensionalised forms of the governing
equations.

That said, you can try normalising the dimensions of your domain. For
example, let's say your domain is 100km by 20km, with height values ranging
from 0m to 10m. Try scaling them down by a factor 1000 to 100m by 20m, with
height ranging from 0 to 0.01m. Do that for your boundary conditions as
well. You can play around with the scale until the solution does not
diverge. Either that or you just reduce your terrain size and compute for a
smaller region.

**Comment by ZhaoMY93** [comment]
Dear,


Thank you very much for your advice, it's very helpful! I'll try reducing the size and see how it goes!


Yours sincerely,

Ray




From: Winston Ong ***@***.***>
Date: 2024-12-22 23:34:24
To: obdwinston/Incompressible-Flow ***@***.***>
Cc: ZhaoMY93 ***@***.***>,Author ***@***.***>
Subject: Re: [obdwinston/Incompressible-Flow] where can i get the swe solver (Issue #1)

Yes, it is very possible that the scale of your terrain may cause the
 solution to diverge. Based on personal experience, that may often be the
 case (assuming code has no errors), not just for this shallow flow solver.
 Hence, it's important that the scales are appropriate. Usually this
 corresponds to working with non-dimensionalised forms of the governing
 equations.

 That said, you can try normalising the dimensions of your domain. For
 example, let's say your domain is 100km by 20km, with height values ranging
 from 0m to 10m. Try scaling them down by a factor 1000 to 100m by 20m, with
 height ranging from 0 to 0.01m. Do that for your boundary conditions as
 well. You can play around with the scale until the solution does not
 diverge. Either that or you just reduce your terrain size and compute for a
 smaller region.

**Comment by obdwinston** [comment]
No problem, keep me posted!

**Comment by ZhaoMY93** [comment]
Dear,


Hello! I remember there was a page of formulas for the shallow water model that you provided, which was extremely clear. If possible, may I request to review it again?


Yours sincerely,

Ray




发件人：Winston Ong ***@***.***>
发送日期：2024-12-23 14:52:15
收件人：obdwinston/Incompressible-Flow ***@***.***>
抄送人：ZhaoMY93 ***@***.***>,Author ***@***.***>
主题：Re: [obdwinston/Incompressible-Flow] where can i get the swe solver (Issue #1)

No problem, keep me posted!

**Comment by obdwinston** [comment]
Sure, please see attached.

**Comment by obdwinston** [comment]
![398544974-51de5641-014a-46a1-bd25-f917eb507bc5](https://github.com/user-attachments/assets/8b30328c-a138-4f9a-a135-65bc87d26074)

**Comment by ZhaoMY93** [comment]
Thank you very much! I get it!






发件人：Winston Ong ***@***.***>
发送日期：2024-12-25 16:44:45
收件人：obdwinston/Incompressible-Flow ***@***.***>
抄送人：ZhaoMY93 ***@***.***>,Author ***@***.***>
主题：Re: [obdwinston/Incompressible-Flow] where can i get the swe solver (Issue #1)

398544974-51de5641-014a-46a1-bd25-f917eb507bc5.png (view on web)
